### DISCUSSION TITLE
Combi-boiler support

### DISCUSSION THREAD

**Discussion by nikshriv** [nikshriv/hass_navien_water_heater#19] [Announcements]
Starting a discussion to get combo units supported…

**Comment by FormerLurker** [comment]
Life has been happening, so I have been unable to work on this much recently.  However, I eill be free this weekend.  I ran into a few issues with the conversions for my unit that further complicated the conversion routine in the navienapi.py file.  To solve this, I created a NavienUnit.py file, some classes/enums, and added the ability to configure each unit separately by model.  Each unit configuration is responsible for converting (via a helper NavienConversions file) units to and from metric/us.  You can also pick which sensors are valid per unit.  I am going to ensure that if multiple boilers/water heaters are on a single channel that all of the entities will be added, and that multiple channels have only one set of controls for supply/DHW/Recirculation.

I hope all this will simplify making changes to the API in the future, and supporting new sensors/units as they become available.  It will just be a matter of adding a config file for the unit, or being ok with the default (all sensors and all controls visible in the default).

Anyway, it is all simpler than it sounds.  I will push to my fork when I have something working and you can check it out.  I have learned a TON from your source (it is my first work on HA), especially regarding the API itself.  V clever, and very cool!!!

**Comment by ninjamojo** [comment]
Ok, so I ran into a snag.  I decided that editing the code and uploading it manually to HA was not the best long term solution, so I got visual studio code setup to run HA via docker (per (these instructions}[https://developers.home-assistant.io/docs/development_environment/]).  HA loads and I can debug.  However, after setting the polling frequency I'm getting an 'ssl handshake error' in navien_api.py on line 150 ( ```await self.loop.run_in_executor(None,self.client.connect)```).  Perhaps the port being used is somehow blocked by docker?  I do get the correct unit being returned from the service initially, it just appears as though the MQTT connect function isn't working.

Not sure what is up, because it seems to run fine on my live HA instance, so I'm guessing it's a dev environment thing.  Any thoughts you might have would be appreciated, as I'm kind of stuck.

**Comment by Wayfinder624** [comment]
Wondering if anyone made progress on Combi-boiler support?

**Comment by FormerLurker** [comment]
Sorry, that was the wrong account.  reposting:

Well, I was making a lot of progress, but had a ton of trouble after I started attempting to develop remotely, and I quit trying. I was getting errors communicating with the API that I just couldn't resolve, and I was unable to determine why. However, recently I am actually back on site here, and might be able to take another look at it. I have one other open source project that needs some attention, then I might take another stab at it..

**Comment by Wayfinder624** [comment]
Thanks for the reply. I am mostly interested in getting the Heating data added.
If I get some time I'll also start looking at the code.

**Comment by plipof** [comment]
I am very interested in combi-bioler functionality to create a custom radiant floor heating controller and outdoor reset functionality.

**Comment by pscheidlerSr** [comment]
This discussion has been dormant for two years.  However, I just had a Navien combi condensing boiler installed, and I was wondering if there was any residual interest in a Navien combi boiler integration.  This integration would be the prerequisite for tuning the Navien boiler to max efficiency.  

The motivation for this note is to encourage the development of a Navien Combi Condensing Boiler integration.  I have done preliminary investigations, and the learning curve for Home Assistant integration is steep.  Since others have traversed it, I want to encourage them to develop a Home Assistant Navien Combi Boiler integration.  

There are economic benefits from the proper tuning of the combi boiler. In my case, the savings would be several hundred dollars a year.  Again, this integration would be the prerequisite for tuning the Navien boiler to max efficiency.  
This note of encouragement presents three sections

1.	Navien Combi Condensing Physics
2.	Modification of Water Heater Integration to generate a Combi Condensing Boiler integration
3.	Home Assistant, Navien Combi Boiler Control, and Efficiency.

**Navien Combi Boiler Condensing Physics**
The hot water used for household heating is heated in two stages.  In the first stage, water is heated by burning natural gas.  In the second stage, energy for heating is extracted from the flue gas generated in the first stage.   
The equation describing this first stage is:  
                          CH4 + O2 -> H2O(hot vapor)  + CO2 + combustion energy
Natural gas (CH4) and oxygen (O2) are burned to generate water (H2O), carbon dioxide (CO2) and energy.  The energy is used to warm the water.  The hot water vapor is passed to a heat exchanger where the vapor is condensed to water and energy is released.
The equation describing the second stage is  
                         H2O(hot vapor)   -> H2O(liquid) + condensation energy

When the hot water vapor is condensed to liquid water energy is released.  This energy is used to warm the water used for household heating.  The temperature at which water vapor condenses, is crucial for efficient operation. For natural gas combi boilers, this temperature is 130F. The proper management of the condensation process is essential to condensing boiler efficiency.  This management will be presented below.

There is a wonderful presentation of combi boiler condensing physics at 
       https://www.reddit.com/r/hvacadvice/comments/1gzr4y3/comment/lyzhuz4/

**Modification of Water Heater Integration to generate a Combi Condensing Boiler integration**

@ Nikshriv has written a wonderful integration presenting data and control for the Navien water heater.  This integration should serve as the platform form for a condensing boiler integration.

The water heater integration presents water heater data and a control for the water temperature.  A combi condensing boiler integration would present additional boiler data and a control for the boiler water temperature.

The navilink_api (Nikshriv) test application presents the Json data available for the combi boiler.  The data items of interest are 'currentInletTemp': 77, 'currentSupplyTemp': 78, 'currentReturnTemp': 77.  The control item that needs to be added is  'heatSettingTemp': 180.  More analysis will probably add additional data items.

The addition of these data and control items to the existing Hot Water integration will require competence in HA integration development.  There is a non-trivial learning curve to acquire this.  The motivation for this post is to encourage someone with this knowledge to pursue this development activity.

**Home Assistant, Navien Combi Boiler Control, and Efficiency**
Once a combi boiler integration is available, the fun begins.  

Hydronic heating systems use a boiler to heat water that is used as the heat transfer medium. Water enters a heating circuit at a high temperature and is cooled as it provides household heating.  

In a condensing boiler, the temperature of the heating water must be adjusted so the temperature of the returned water is less than 130 degrees.  If the returned water temperature from the heating circuit is greater than 130 degrees, the water vapor in the flue gas will not condense, and the boiler runs inefficiently.  When the returned water temperature is less than 130 degrees, the water vapor in the flue gas condenses, energy is extracted to heat the returned water and the boiler runs efficiently. 

But, water entering the heating circuit at a lower temperature might not provide the energy needed for heating.  Thus, there may be a trade between efficiency and comfort. 

The variables that determine the difference between the temperature of water entering and exiting a heating circuit include:
•	The outdoor temperature
•	The length of the heating circuit
•	The heat loss of the building
•	Number of zones calling for heat
•	Time of day
•	????
These issues are site specific and can be addressed in Home Assistant.  Addressing them can yield real economic value.  A combi boiler integration is a prerequisite.
